### PR TITLE
Fix Uncaught TypeError with some shifts (#5921)

### PR DESCRIPTION
**Pull request by BafS**
Fix https://github.com/vimeo/psalm/issues/5921

### CODE REVIEW

**Review by orklah — APPROVED** [review]

**Review by weirdan — COMMENTED** [review]
Mind adding a test or two for boundary conditions? E.g. PHP produces float for `2**63` (https://3v4l.org/iWb5G) which might result in a failed assertion.

File: src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/NonDivArithmeticOpAnalyzer.php [hunk 1]
```diff
@@ -273,6 +275,19 @@ public static function analyze(
...
 
+    /**
+     * @param int|float $result
+     */
+    private static function getNumericalType($result): Type\Union
+    {
+        if ($result <= PHP_INT_MAX && $result >= PHP_INT_MIN) {
+            assert(is_int($result));
+            return Type::getInt(false, $result);
+        }
```
Comment (weirdan) [inline src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/NonDivArithmeticOpAnalyzer.php:286]:
```suggestion
        if (is_int($result)) {
            return Type::getInt(false, $result);
        }
```
I think it's the most robust way to check that.

**Review by weirdan — CHANGES_REQUESTED** [review]

File: src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/NonDivArithmeticOpAnalyzer.php (hunk 1 above)
Comment (orklah) [reply src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/NonDivArithmeticOpAnalyzer.php:286]: Yep, I should have gone there from the start!

Comment (BafS) [reply src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/NonDivArithmeticOpAnalyzer.php]: Yes thanks !

**Comment by BafS** [comment]
Clean the code and added an assert for `2 ** 63`

File: tests/BinaryOperationTest.php [hunk 2]
```diff
@@ -532,6 +532,20 @@ function foo($a, $b): void {
                     '$a' => 'float'
                 ],
             ],
+            'IntOverflowPowSub' => [
```
Comment (weirdan) [inline tests/BinaryOperationTest.php:535]:
```suggestion
            'IntOverflowPlus' => [
                '<?php
                    $a = 2**62 - 1 + 2**62; 
                    $b = 2**62 + 2**62 - 1; // plus results in a float',
                'assertions' => [
                    '$a' => 'int',
                    '$b' => 'float',
                ],
            ],
            'IntOverflowPowSub' => [
```

**Review by weirdan — APPROVED** [review]

**Comment by weirdan** [comment]
Thanks!
